File: tests/test_performance.py
import time
import random
import psutil
from stats import mean, median

process = psutil.Process()

def get_memory_usage():
    """Returns the memory usage in MB."""
    return process.memory_info().rss / 1024 ** 2  # Convert bytes to MB

def test_mean_performance():
    numbers = [random.random() for _ in range(10_000_000)]  # Use 10 million elements

    # Measure initial memory
    initial_memory = get_memory_usage()

    # Start timing
    start_time = time.time()
    mean(numbers)
    duration = time.time() - start_time

    # Measure final memory
    final_memory = get_memory_usage()

    print(f"Time taken for mean calculation: {duration:.6f} seconds")
    print(f"Initial memory usage: {initial_memory:.2f} MB")
    print(f"Final memory usage: {final_memory:.2f} MB")
    print(f"Memory consumed by mean calculation: {final_memory - initial_memory:.2f} MB")

def test_median_performance():
    numbers = [random.random() for _ in range(10_000_000)]  # Use 10 million elements

    # Measure initial memory
    initial_memory = get_memory_usage()

    # Start timing
    start_time = time.time()
    median(numbers)
    duration = time.time() - start_time

    # Measure final memory
    final_memory = get_memory_usage()

    print(f"Time taken for median calculation: {duration:.6f} seconds")
    print(f"Initial memory usage: {initial_memory:.2f} MB")
    print(f"Final memory usage: {final_memory:.2f} MB")
    print(f"Memory consumed by median calculation: {final_memory - initial_memory:.2f} MB")

if __name__ == "__main__":
    test_mean_performance()
    test_median_performance()

File: tests/test_performance.rs
use std::fs::OpenOptions;
use std::io::Write;
use std::time::Instant;
use sysinfo::{System, SystemExt};
use yijia_ids706_mini_proj8::stats::{mean, median};

fn write_to_md(content: &str) {
    let mut file = OpenOptions::new()
        .create(true)
        .write(true)
        .append(true)
        .open("rust_performance.md")
        .expect("Unable to open rust_performance.md");
    writeln!(file, "{}", content).expect("Unable to write data");
}

#[test]
fn test_mean_performance() {
    let numbers = vec![1.0; 10_000_000]; // Use 10 million elements

    // Initialize the system for memory tracking
    let mut sys = System::new_all();
    sys.refresh_memory();
    let memory_before = sys.used_memory();

    let start = Instant::now();
    mean(&numbers);
    let duration = start.elapsed();

    // Capture memory after execution
    sys.refresh_memory();
    let memory_after = sys.used_memory();
    let memory_used = memory_after as i64 - memory_before as i64;

    let mean_report = format!(
        "### Mean Calculation Performance\n\n\
        - Time taken: {:?}\n\
        - Initial memory usage: {} KB\n\
        - Final memory usage: {} KB\n\
        - Memory consumed: {} KB\n\n",
        duration, memory_before, memory_after, memory_used
    );
    write_to_md(&mean_report);

    println!("Rust - Time taken for mean calculation: {:?}", duration);
    println!(
        "Rust - Memory usage for mean calculation: {} KB",
        memory_used
    );
}

#[test]
fn test_median_performance() {
    let mut numbers = vec![1.0; 10_000_000];

    // Initialize the system for memory tracking
    let mut sys = System::new_all();
    sys.refresh_memory();
    let memory_before = sys.used_memory();

    let start = Instant::now();
    median(&mut numbers);
    let duration = start.elapsed();

    // Capture memory after execution
    sys.refresh_memory();
    let memory_after = sys.used_memory();
    let memory_used = memory_after as i64 - memory_before as i64;

    let median_report = format!(
        "### Median Calculation Performance\n\n\
        - Time taken: {:?}\n\
        - Initial memory usage: {} KB\n\
        - Final memory usage: {} KB\n\
        - Memory consumed: {} KB\n\n",
        duration, memory_before, memory_after, memory_used
    );
    write_to_md(&median_report);

    println!("Rust - Time taken for median calculation: {:?}", duration);
    println!(
        "Rust - Memory usage for median calculation: {} KB",
        memory_used
    );
}
